# PART: 1Segmenting and Clustering Neighborhoods in Toronto

In [4]:
!pip install lxml

     |████████████████████████████████| 5.8MB 6.9MB/s eta 0:00:01


In [1]:
# Import data from wikipedia

import pandas as pd
import numpy as np

link = "https://en.wikipedia.org/wiki/List_of_postal_codes_of_Canada:_M"
pd.set_option('display.max_colwidth', -1)

In [2]:
# Use Pandas read_html to get the table from the WikiPedia page
# The table we are interested in is the first table on the page

df = pd.read_html(link, header=0)[0]

In [3]:
df.head(10)

Postcode           Borough     Neighbourhood
0  M1A      Not assigned      Not assigned    
1  M2A      Not assigned      Not assigned    
2  M3A      North York        Parkwoods       
3  M4A      North York        Victoria Village
4  M5A      Downtown Toronto  Harbourfront    
5  M6A      North York        Lawrence Heights
6  M6A      North York        Lawrence Manor  
7  M7A      Downtown Toronto  Queen's Park    
8  M8A      Not assigned      Not assigned    
9  M9A      Queen's Park      Not assigned

In [4]:
df.shape

(287, 3)

### Data cleaning

In [5]:
# drop Borough colomns not assigned

df = df[df.Borough != 'Not assigned']

# If a cell has a borough but a Not assigned neighborhood, then the neighborhood will be the same as the borough

df.loc[df['Neighbourhood'] == 'Not assigned', 'Neighbourhood'] = df['Borough']

In [6]:
df.shape

(210, 3)

After cleaning "Not assigned" values the dataframe reduced from 287 to 210 rows.

### Group by Postal Code and Borough

In [7]:
# Gruop neighborhood wiht the same Post code and produce a new list of neighbourhoods separed by comma.

dfs= pd.DataFrame(df.groupby(
    ['Postcode', 'Borough'])['Neighbourhood'].apply(
    lambda x: ', '.join(x))).reset_index()

In [8]:
dfs.head(10)

Postcode      Borough                                    Neighbourhood
0  M1B      Scarborough  Rouge, Malvern                                 
1  M1C      Scarborough  Highland Creek, Rouge Hill, Port Union         
2  M1E      Scarborough  Guildwood, Morningside, West Hill              
3  M1G      Scarborough  Woburn                                         
4  M1H      Scarborough  Cedarbrae                                      
5  M1J      Scarborough  Scarborough Village                            
6  M1K      Scarborough  East Birchmount Park, Ionview, Kennedy Park    
7  M1L      Scarborough  Clairlea, Golden Mile, Oakridge                
8  M1M      Scarborough  Cliffcrest, Cliffside, Scarborough Village West
9  M1N      Scarborough  Birch Cliff, Cliffside West

In [9]:
dfs.shape

(103, 3)

The dataframe continue with the same amount of rows 103.

#  PART2: latitude and the longitude coordinates of each neighborhood

We will built a dataframe of the postal code of each neighborhood along with the borough name and neighborhood name

In [10]:
# Install geocoder if required
!conda install -c conda-forge geocoder --yes

Solving environment: done


==> WARNING: A newer version of conda exists. <==
  current version: 4.5.11
  latest version: 4.8.2

Please update conda by running

    $ conda update -n base -c defaults conda



## Package Plan ##

  environment location: /home/jupyterlab/conda/envs/python

  added / updated specs: 
    - geocoder


The following packages will be downloaded:

    package                    |            build
    ---------------------------|-----------------
    certifi-2019.11.28         |           py36_0         149 KB  conda-forge
    scikit-learn-0.20.1        |   py36h22eb022_0         5.7 MB
    liblapack-3.8.0            |      11_openblas          10 KB  conda-forge
    numpy-1.18.1               |   py36h95a1406_0         5.2 MB  conda-forge
    liblapacke-3.8.0           |      11_openblas          10 KB  conda-forge
    libopenblas-0.3.6          |       h5a2b251_2         7.7 MB
    scipy-1.4.1                |   py36h921218d_0        18.9 MB  conda-forge
    

In [11]:
# Lets create a copy 2 from part 1
c2 = dfs

In [13]:
# Lets create a copy 3 in case the previus loop doenst work
c3 = dfs

In [14]:
# create a dataframe data with the data that we will directly download for the latitude and longitude
data = pd.read_csv('https://cocl.us/Geospatial_data')

In [15]:
data.head()

Postal Code   Latitude  Longitude
0  M1B         43.806686 -79.194353
1  M1C         43.784535 -79.160497
2  M1E         43.763573 -79.188711
3  M1G         43.770992 -79.216917
4  M1H         43.773136 -79.239476

### Join the 2 dataframe copy 3 and data

In [16]:
c3 = pd.concat([c3, data], axis=1, join='inner')
c3.drop('Postal Code', axis=1, inplace=True)

In [17]:
c3.head()

Postcode      Borough                           Neighbourhood   Latitude  \
0  M1B      Scarborough  Rouge, Malvern                          43.806686   
1  M1C      Scarborough  Highland Creek, Rouge Hill, Port Union  43.784535   
2  M1E      Scarborough  Guildwood, Morningside, West Hill       43.763573   
3  M1G      Scarborough  Woburn                                  43.770992   
4  M1H      Scarborough  Cedarbrae                               43.773136   

   Longitude  
0 -79.194353  
1 -79.160497  
2 -79.188711  
3 -79.216917  
4 -79.239476

In [18]:
c3.shape

(103, 5)

# PART 3: Explore and cluster the neighborhoods in Toronto

In [20]:
# Importing libraries

import requests # library to handle requests

!conda install -c conda-forge geopy --yes # we imported in Part:2
from geopy.geocoders import Nominatim # module to convert an address into latitude and longitude values

# libraries for displaying images
from IPython.display import Image 
from IPython.core.display import HTML 

# Tranform JSON file into a pandas dataframe
from pandas.io.json import json_normalize 


!conda install -c conda-forge folium=0.5.0 --yes
import folium # plotting library

print('Folium installed')
print('Libraries imported.')

Solving environment: done


==> WARNING: A newer version of conda exists. <==
  current version: 4.5.11
  latest version: 4.8.2

Please update conda by running

    $ conda update -n base -c defaults conda



## Package Plan ##

  environment location: /home/jupyterlab/conda/envs/python

  added / updated specs: 
    - geopy


The following packages will be downloaded:

    package                    |            build
    ---------------------------|-----------------
    geopy-1.21.0               |             py_0          58 KB  conda-forge
    geographiclib-1.50         |             py_0          34 KB  conda-forge
    ------------------------------------------------------------
                                           Total:          92 KB

The following NEW packages will be INSTALLED:

    geographiclib: 1.50-py_0   conda-forge
    geopy:         1.21.0-py_0 conda-forge


geopy-1.21.0         | 58 KB     | ##################################### | 100% 
geographiclib-1.50   | 3

we will work with only boroughs that contain the word Toronto

In [21]:
# new dataframe Toronto droping what we dont need from last copy 3
toronto = c3[c3['Borough'].str.contains("Toronto", case=False)]

In [22]:
toronto.head()

Postcode          Borough                   Neighbourhood   Latitude  \
37  M4E      East Toronto     The Beaches                     43.676357   
41  M4K      East Toronto     The Danforth West, Riverdale    43.679557   
42  M4L      East Toronto     The Beaches West, India Bazaar  43.668999   
43  M4M      East Toronto     Studio District                 43.659526   
44  M4N      Central Toronto  Lawrence Park                   43.728020   

    Longitude  
37 -79.293031  
41 -79.352188  
42 -79.315572  
43 -79.340923  
44 -79.388790

In [23]:
toronto.shape

(39, 5)

####  Foursquare Credentials and Version

In [24]:

CLIENT_ID = 'BWPDO2AOHC31B0LBOZWGDWTO2DW5FS3FZIU4RKDD5IYIMVKG' # your Foursquare ID
CLIENT_SECRET = 'ENGF5W43FNDSC5OTRV4115VJEN42TEGREY33NF12KMV2LLHK' # your Foursquare Secret
VERSION = '20180604'
LIMIT = 100
RADIUS = 1000

### Use geopy library to get the latitude and longitude values of Toronto

In [25]:
# Lets get the latitude and longitude of toronto to draw a map first
address = 'Toronto, Canada'

geolocator = Nominatim(user_agent="ny_explorer")
location = geolocator.geocode(address)
latitude = location.latitude
longitude = location.longitude
print('The geograpical coordinate of Toronto are {}, {}.'.format(latitude, longitude))

The geograpical coordinate of Toronto are 43.653963, -79.387207.


### Create a Map of Toronto with his neighborhoods

In [26]:
## Lets visualise the neighborhood using latitude and longitude values
map_toronto = folium.Map(location=[latitude, longitude], zoom_start=11)

# add markers to map
for lat, lng, label in zip(toronto['Latitude'], toronto['Longitude'], toronto['Borough']):
    label = folium.Popup(label, parse_html=True)
    folium.CircleMarker(
        [lat, lng],
        radius=5,
        popup=label,
        color='blue',
        fill=True,
        fill_color='#3186cc',
        fill_opacity=0.7,
        parse_html=False).add_to(map_toronto)  
    
map_toronto

### Toronto Venues

In [27]:
# function that extracts the category of the venue
def get_category_type(row):
    try:
        categories_list = row['categories']
    except:
        categories_list = row['venue.categories']
        
    if len(categories_list) == 0:
        return None
    else:
        return categories_list[0]['name']

In [28]:
def getNearbyVenues(names, latitudes, longitudes, radius=500):
    
    venues_list=[]
    for name, lat, lng in zip(names, latitudes, longitudes):
            
        # create the API request URL
        url = 'https://api.foursquare.com/v2/venues/explore?&client_id={}&client_secret={}&v={}&ll={},{}&radius={}&limit={}'.format(
            CLIENT_ID, 
            CLIENT_SECRET, 
            VERSION, 
            lat, 
            lng, 
            radius, 
            LIMIT)
            
        # make the GET request
        results = requests.get(url).json()["response"]['groups'][0]['items']
        # return only relevant information for each nearby venue
        venues_list.append([(
            name, 
            lat, 
            lng, 
            v['venue']['name'], 
            v['venue']['location']['lat'], 
            v['venue']['location']['lng'],  
            v['venue']['categories'][0]['name']) for v in results])

    nearby_venues = pd.DataFrame([item for venue_list in venues_list for item in venue_list])
    nearby_venues.columns = ['Neighborhood', 
                  'Neighborhood Latitude', 
                  'Neighborhood Longitude', 
                  'Venue', 
                  'Venue Latitude', 
                  'Venue Longitude', 
                  'Venue Category']
    
    return(nearby_venues)

In [29]:
# Function to sort the venues in descending order.
def return_most_common_venues(row, num_top_venues):
    row_categories = row.iloc[1:]
    row_categories_sorted = row_categories.sort_values(ascending=False)
    return row_categories_sorted.index.values[0:num_top_venues]

In [30]:
toronto_venues = getNearbyVenues(names=toronto['Postcode'],
                                 latitudes=toronto['Latitude'],
                                 longitudes=toronto['Longitude']
                                )

In [31]:
toronto_venues

Neighborhood  Neighborhood Latitude  Neighborhood Longitude  \
0     M4E          43.676357             -79.293031                
1     M4E          43.676357             -79.293031                
2     M4E          43.676357             -79.293031                
3     M4E          43.676357             -79.293031                
4     M4K          43.679557             -79.352188                
...   ...                ...                    ...                
1700  M7Y          43.662744             -79.321558                
1701  M7Y          43.662744             -79.321558                
1702  M7Y          43.662744             -79.321558                
1703  M7Y          43.662744             -79.321558                
1704  M7Y          43.662744             -79.321558                

                                   Venue  Venue Latitude  Venue Longitude  \
0     Glen Manor Ravine                   43.676821      -79.293942         
1     The Big Carrot Natural Food Market  43.678879      -79.297734         
2     Grover Pub and Grub                 43.679181      -79.297215         
3     Upper Beaches                       43.680563      -79.292869         
4     Pantheon                            43.677621      -79.351434         
...        ...                                  ...             ...         
1700  Jonathan Ashbridge Park             43.664702      -79.319898         
1701  The Ten Spot                        43.664815      -79.324213         
1702  Olliffe On Queen                    43.664503      -79.324768         
1703  Greenwood Cigar & Variety           43.664538      -79.325379         
1704  Revolution Recording                43.662561      -79.326940         

         Venue Category  
0     Trail              
1     Health Food Store  
2     Pub                
3     Neighborhood       
4     Greek Restaurant   
...                ...   
1700  Park               
1701  Spa                
1702  Butcher            
1703  Smoke Shop         
1704  Recording Studio   

[1705 rows x 7 columns]

In [32]:
toronto_venues.groupby('Neighborhood').count()

Neighborhood Latitude  Neighborhood Longitude  Venue  \
Neighborhood                                                         
M4E           4                      4                       4       
M4K           42                     42                      42      
M4L           18                     18                      18      
M4M           41                     41                      41      
M4N           4                      4                       4       
M4P           7                      7                       7       
M4R           23                     23                      23      
M4S           35                     35                      35      
M4T           4                      4                       4       
M4V           14                     14                      14      
M4W           4                      4                       4       
M4X           42                     42                      42      
M4Y           81                     81                      81      
M5A           46                     46                      46      
M5B           100                    100                     100     
M5C           100                    100                     100     
M5E           55                     55                      55      
M5G           84                     84                      84      
M5H           100                    100                     100     
M5J           100                    100                     100     
M5K           100                    100                     100     
M5L           100                    100                     100     
M5N           3                      3                       3       
M5P           4                      4                       4       
M5R           21                     21                      21      
M5S           36                     36                      36      
M5T           80                     80                      80      
M5V           16                     16                      16      
M5W           94                     94                      94      
M5X           100                    100                     100     
M6G           18                     18                      18      
M6H           19                     19                      19      
M6J           51                     51                      51      
M6K           24                     24                      24      
M6P           25                     25                      25      
M6R           14                     14                      14      
M6S           40                     40                      40      
M7A           39                     39                      39      
M7Y           17                     17                      17      

              Venue Latitude  Venue Longitude  Venue Category  
Neighborhood                                                   
M4E           4               4                4               
M4K           42              42               42              
M4L           18              18               18              
M4M           41              41               41              
M4N           4               4                4               
M4P           7               7                7               
M4R           23              23               23              
M4S           35              35               35              
M4T           4               4                4               
M4V           14              14               14              
M4W           4               4                4               
M4X           42              42               42              
M4Y           81              81               81              
M5A           46              46               46              
M5B           100             100              100             
M5C           100             100              100     

**Let's find out how many unique categories can be curated from all the returned venues**


In [33]:
print('There are {} uniques categories.'.format(len(toronto_venues['Venue Category'].unique())))

There are 229 uniques categories.


In [34]:
size = toronto_venues.Neighborhood.value_counts()
idx = size[size >= 10].index
toronto_venues_filtered = toronto_venues[toronto_venues.Neighborhood.isin(idx)]
toronto_venues_filtered.shape

(1675, 7)

### Analyze Each Neighborhood


In [35]:
# Convert each category to one hot
toronto_one_hot = pd.get_dummies(toronto_venues[['Venue Category']], prefix="", prefix_sep="")
# Add neighborhood back
toronto_one_hot['Neighborhood'] = toronto_venues['Neighborhood'] 

# move neighborhood column to the first column
toronto_one_hot = toronto_one_hot.set_index('Neighborhood').reset_index()

toronto_one_hot.head()

Neighborhood  Afghan Restaurant  Airport  Airport Food Court  Airport Gate  \
0  M4E          0                  0        0                   0              
1  M4E          0                  0        0                   0              
2  M4E          0                  0        0                   0              
3  M4E          0                  0        0                   0              
4  M4K          0                  0        0                   0              

   Airport Lounge  Airport Service  Airport Terminal  American Restaurant  \
0  0               0                0                 0                     
1  0               0                0                 0                     
2  0               0                0                 0                     
3  0               0                0                 0                     
4  0               0                0                 0                     

   Antique Shop  ...  Theme Restaurant  Toy / Game Store  Trail  \
0  0             ...  0                 0                 1       
1  0             ...  0                 0                 0       
2  0             ...  0                 0                 0       
3  0             ...  0                 0                 0       
4  0             ...  0                 0                 0       

   Train Station  Vegetarian / Vegan Restaurant  Video Game Store  \
0  0              0                              0                  
1  0              0                              0                  
2  0              0                              0                  
3  0              0                              0                  
4  0              0                              0                  

   Vietnamese Restaurant  Wine Bar  Women's Store  Yoga Studio  
0  0                      0         0              0            
1  0                      0         0              0            
2  0                      0         0              0            
3  0                      0         0              0            
4  0                      0         0              0            

[5 rows x 229 columns]

In [36]:
toronto_one_hot.shape

(1705, 229)

### Best 5 venues from each neighborhood

In [37]:


# Next, let's group rows by neighborhood and by taking the mean of the frequency of occurrence of each category¶

toronto_grouped = toronto_one_hot.groupby('Neighborhood').mean().reset_index()
toronto_grouped.head()
num_top_venues = 5


# Lets print each neighboorhood with the top 5 most common venue

for hood in toronto_grouped['Neighborhood']:
    print("----"+hood+"----")
    temp = toronto_grouped[toronto_grouped['Neighborhood'] == hood].T.reset_index()
    temp.columns = ['venue','freq']
    temp = temp.iloc[1:]
    temp['freq'] = temp['freq'].astype(float)
    temp = temp.round({'freq': 2})
    print(temp.sort_values('freq', ascending=False).reset_index(drop=True).head(num_top_venues))
    print('\n')

----M4E----
               venue  freq
0  Health Food Store  0.25
1  Trail              0.25
2  Pub                0.25
3  Afghan Restaurant  0.00
4  Movie Theater      0.00


----M4K----
                venue  freq
0  Greek Restaurant    0.24
1  Coffee Shop         0.10
2  Italian Restaurant  0.07
3  Restaurant          0.05
4  Ice Cream Shop      0.05


----M4L----
              venue  freq
0  Park              0.11
1  Brewery           0.06
2  Steakhouse        0.06
3  Sushi Restaurant  0.06
4  Sandwich Place    0.06


----M4M----
                 venue  freq
0  Café                 0.10
1  Coffee Shop          0.07
2  Brewery              0.05
3  American Restaurant  0.05
4  Bakery               0.05


----M4N----
               venue  freq
0  Park               0.25
1  Lake               0.25
2  Bus Line           0.25
3  Swim School        0.25
4  Afghan Restaurant  0.00


----M4P----
               venue  freq
0  Gym                0.14
1  Park               0.14
2  Breakfast Sp

**Let's put that into a pandas dataframe**

In [38]:
### First, let's write a function to sort the venues in descending order.

def return_most_common_venues(row, num_top_venues):
    row_categories = row.iloc[1:]
    row_categories_sorted = row_categories.sort_values(ascending=False)
    
    return row_categories_sorted.index.values[0:num_top_venues]


In [39]:

num_top_venues = 10

indicators = ['st', 'nd', 'rd']

# create columns according to number of top venues
columns = ['Neighborhood']
for ind in np.arange(num_top_venues):
    try:
        columns.append('{}{} Most Common Venue'.format(ind+1, indicators[ind]))
    except:
        columns.append('{}th Most Common Venue'.format(ind+1))

# create a new dataframe
neighborhoods_venues_sorted = pd.DataFrame(columns=columns)
neighborhoods_venues_sorted['Neighborhood'] = toronto_grouped['Neighborhood']

for ind in np.arange(toronto_grouped.shape[0]):
    neighborhoods_venues_sorted.iloc[ind, 1:] = return_most_common_venues(toronto_grouped.iloc[ind, :], num_top_venues)

print(neighborhoods_venues_sorted.shape)
neighborhoods_venues_sorted.head()

(39, 11)


Neighborhood 1st Most Common Venue 2nd Most Common Venue  \
0  M4E          Trail                 Health Food Store      
1  M4K          Greek Restaurant      Coffee Shop            
2  M4L          Park                  Board Shop             
3  M4M          Café                  Coffee Shop            
4  M4N          Lake                  Swim School            

  3rd Most Common Venue 4th Most Common Venue   5th Most Common Venue  \
0  Pub                   Yoga Studio           Deli / Bodega            
1  Italian Restaurant    Ice Cream Shop        Furniture / Home Store   
2  Sushi Restaurant      Sandwich Place        Brewery                  
3  Gastropub             Bakery                Italian Restaurant       
4  Bus Line              Park                  General Entertainment    

  6th Most Common Venue        7th Most Common Venue 8th Most Common Venue  \
0  Electronics Store     Eastern European Restaurant  Dumpling Restaurant    
1  Restaurant            Bubble Tea Shop              Grocery Store          
2  Liquor Store          Burger Joint                 Italian Restaurant     
3  Brewery               American Restaurant          Yoga Studio            
4  Deli / Bodega         Eastern European Restaurant  Dumpling Restaurant    

  9th Most Common Venue 10th Most Common Venue  
0  Donut Shop            Doner Restaurant       
1  Pub                   Pizza Place            
2  Burrito Place         Fast Food Restaurant   
3  Bookstore             Sandwich Place         
4  Donut Shop            Doner Restaurant

### Cluster Neighborhoods

In [40]:
# set number of clusters
kclusters = 5

toronto_grouped_clustering = toronto_grouped.drop('Neighborhood', 1)

# run k-means clustering
kmeans = KMeans(n_clusters=kclusters, random_state=0).fit(toronto_grouped_clustering)

# check cluster labels generated for each row in the dataframe
kmeans.labels_[0:10]

NameError: name 'KMeans' is not defined

In [ ]:
# define the dataframe columns
column_names = ['PostalCode', 'Borough', 'Neighborhood']

# instantiate the dataframe
neighborhoods = pd.DataFrame(columns=column_names)
# Lets check which are the rows that have Neighborhood as "Not assigned"
neighborhoods[neighborhoods['Neighborhood'] == 'Not assigned']

In [ ]:

# add clustering labels
neighborhoods_venues_sorted = neighborhoods_venues_sorted.drop('Cluster Labels', 1)
neighborhoods_venues_sorted.insert(0, 'Cluster Labels', kmeans.labels_)

toronto_merged = neigh_latlong

# merge toronto_grouped with toronto_data to add latitude/longitude for each neighborhood
toronto_merged = toronto_merged.join(neighborhoods_venues_sorted.set_index('Neighborhood'), on='Neighborhood')

toronto_merged # check the last columns!